### PR TITLE
schedule: refactor region scatter

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?
ref https://github.com/tikv/pd/issues/3450
Also the code in region scatter is too complex to maintain.

### What is changed and how it works?
refactor the region scatter to make it simple and fix the reference bug.

After this pr, the result was like following: 

```sh
mysql> create table t(id int);
Query OK, 0 rows affected (0.09 sec)

mysql> SPLIT TABLE t BETWEEN (0) AND (1000000000) REGIONS 200;                                +--------------------+----------------------+
| TOTAL_SPLIT_REGION | SCATTER_FINISH_RATIO |
+--------------------+----------------------+
|                199 |                    1 |
+--------------------+----------------------+
1 row in set (11.89 sec)

mysql> select count(s.region_id) cnt, s.index_name, p.store_id from INFORMATION_SCHEMA.TIKV_REGION_STATUS s join INFORMATION_SCHEMA.tikv_region_peers p on s.region_id = p.region_id where s.table_name = 't' and p.is_leader = 0 group by index_name, p.store_id order by index_name,cnt desc;
+-----+------------+----------+
| cnt | index_name | store_id |
+-----+------------+----------+
|  70 | NULL       |        1 |
|  67 | NULL       |        4 |
|  67 | NULL       |        6 |
|  66 | NULL       |        7 |
|  65 | NULL       |        5 |
|  65 | NULL       |        8 |
+-----+------------+----------+
6 rows in set (0.02 sec)
```

For the detail test result, please see:
https://docs.google.com/document/d/1cGZU2DCm0UuIueqAQP8TGqTivyavCvB_QiwIzqvP_rA/edit?usp=sharing

### Check List

Tests

- Unit test

### Release note

* No release note

### CODE REVIEW

File: server/schedule/filter/filters.go [hunk 1]
```diff
@@ -385,6 +388,8 @@ func (f *StoreStateFilter) anyConditionMatch(typ int, opt *config.PersistOptions
 	case regionTarget:
 		funcs = []conditionFunc{f.isTombstone, f.isOffline, f.isDown, f.isDisconnected, f.isBusy,
 			f.exceedAddLimit, f.tooManySnapshots, f.tooManyPendingPeers}
+	case scatterRegionTarget:
+		funcs = []conditionFunc{f.isTombstone, f.isOffline, f.isDown, f.isDisconnected}
```
Comment (Yisaer) [inline server/schedule/filter/filters.go:392]: If the target store is chosen by `scatter`, we can ignore some limitation like `exceedAddLimit`

**Comment by nolouch** [comment]
Could you describe the root cause in #3450?

**Comment by Yisaer** [comment]
> Could you describe the root cause in #3450?

updated.

File: server/schedule/region_scatterer.go [hunk 2]
```diff
@@ -41,81 +40,34 @@ var gcInterval = time.Minute
...
-
-func (s *selectedStores) getGroupDistributionOrDefault(group string) map[uint64]uint64 {
-	if result, ok := s.getGroupDistribution(group); ok {
-		return result
-	}
-	return make(map[uint64]uint64)
-}
-
-func (s *selectedStores) put(id uint64, group string) bool {
+// Put ...
```
Comment (JmPotato) [inline server/schedule/region_scatterer.go:53]: Finish up the comments here?

File: server/schedule/region_scatterer.go [hunk 3]
```diff
@@ -41,81 +40,34 @@ var gcInterval = time.Minute
...
-		return
-	}
-	s.stores.Clear()
-}
-
-func (s *selectedStores) get(id uint64, group string) uint64 {
-	s.mu.Lock()
-	defer s.mu.Unlock()
-	distribution, ok := s.getGroupDistribution(group)
+// Get ...
```
Comment (JmPotato) [inline server/schedule/region_scatterer.go:66]: Ditto.

File: server/schedule/region_scatterer.go [hunk 4]
```diff
@@ -126,19 +78,18 @@ func (s *selectedStores) get(id uint64, group string) uint64 {
...
-	defer s.mu.Unlock()
-	if !s.checkExist {
-		return nil
-	}
-	cloned := make(map[uint64]struct{})
-	if groupPlaced, ok := s.getStore(group); ok {
-		for id := range groupPlaced {
-			cloned[id] = struct{}{}
-		}
+// GetGroupDistribution ...
```
Comment (JmPotato) [inline server/schedule/region_scatterer.go:81]: Ditto.

**Review by JmPotato — COMMENTED** [review]
I still need to digest the rest of the codes, so I'll continue the review later.

**Review by longfangsong — APPROVED** [review]
/lgtm

File: server/schedule/region_scatterer.go [hunk 5]
```diff
@@ -126,19 +78,18 @@ func (s *selectedStores) get(id uint64, group string) uint64 {
...
-			cloned[id] = struct{}{}
-		}
+// GetGroupDistribution get distribution group by `group`
+func (s *selectedStores) GetGroupDistribution(group string) (map[uint64]uint64, bool) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	return s.getDistributionByGroup(group)
+}
+
+func (s *selectedStores) getDistributionByGroup(group string) (map[uint64]uint64, bool) {
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply server/schedule/region_scatterer.go]: updated.

File: server/schedule/region_scatterer.go [hunk 6]
```diff
@@ -172,8 +124,8 @@ func newEngineContext(ctx context.Context, filters ...filter.Filter) engineConte
 	filters = append(filters, &filter.StoreStateFilter{ActionScope: regionScatterName})
```
Comment (nolouch) [inline server/schedule/region_scatterer.go:124]: Need set `StoreStateFilter` to true?

Comment (Yisaer) [reply server/schedule/region_scatterer.go]: updated.

File: server/schedule/region_scatterer.go [hunk 7]
```diff
@@ -346,74 +286,60 @@ func (r *RegionScatterer) scatterRegion(region *core.RegionInfo, group string) *
...
+	}
+	return candidates
+}
+
+func (r *RegionScatterer) selectStore(group string, peer *metapb.Peer, candidates []uint64, context engineContext) *metapb.Peer {
+	if len(candidates) < 1 {
+		return peer
+	}
+	var newPeer *metapb.Peer
+	minCount := uint64(math.MaxUint64)
```
Comment (nolouch) [inline server/schedule/region_scatterer.go:328]: should we need to compare with the count of source peer's store? and add a test case `TestFourStores ` with 1000 regions to check it.

Comment (Yisaer) [reply server/schedule/region_scatterer.go:328]: updated.

**Review by nolouch — APPROVED** [review]
lgtm

**Review by HunDunDM — APPROVED** [review]

**Pull request merged**